Avoid deref of Ref in copy constructor of PackedCandidate

Dr15Jones opened the pull request:
#### PR description:

Previously if the vertex had not been read from the original, a call to a copy constructor would cause the deref to happen.

The copy constructor was being called by ROOT during read back from a file. This meant while we were reading the branch holding the PackedCandidate we were trigger reading the branch holding the Vertexes.

#### PR validation:

The code compiles and the extended unit test passes.

Dr15Jones committed:
Avoid deref of Ref in copy constructor of PackedCandidate

Previously if the vertex had not been read from the original, a
call to a copy constructor would cause the deref to happen.

Dr15Jones commented:
Please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32964/21199

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

DataFormats/PatCandidates


@perrotta, @jpata, @slava77 can you please review it and eventually sign? Thanks.
@gpetruc, @cbernet, @gouskos, @rovere, @hatakeyamak this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10b87b/13016/summary.html
**COMMIT**: 22f5521804d2e9a1ffee48aeefecca1bef226295
**CMSSW**: CMSSW_11_3_X_2021-02-22-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32964/13016/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2750983
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2750954
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

slava77 commented:
> Previously if the vertex had not been read from the original, a call to a copy constructor would cause the deref to happen.
> 
> The copy constructor was being called by ROOT during read back from a file. This meant while we were reading the branch holding the PackedCandidate we were trigger reading the branch holding the Vertexes.

is there a chance that  a `    PackedCandidate(PackedCandidate &&iOther)` is called or any of the assignment operators?

slava77 commented on DataFormats/PatCandidates/interface/PackedCandidate.h:
```diff
@@ -204,12 +204,12 @@ namespace pat {
-          deta_(iOther.deta_),
-          dtrkpt_(iOther.dtrkpt_),
+          vertex_((iOther.vertex_ ? new Point(iOther.vertex()) : nullptr)),
+          dxy_(vertex_ ? iOther.dxy_ : 0),
+          dz_(vertex_ ? iOther.dz_ : 0),
+          dphi_(vertex_ ? iOther.dphi_ : 0),
+          deta_(vertex_ ? iOther.deta_ : 0),
+          dtrkpt_(vertex_ ? iOther.dtrkpt_ : 0),
```
this doesn't look right; having a null vertex does not imply zero values in these parameters
https://github.com/cms-sw/cmssw/blob/22f5521804d2e9a1ffee48aeefecca1bef226295/DataFormats/PatCandidates/src/PackedCandidate.cc#L122-L132
I think that the unpacking of parameters not dependent on the value of the `vertexRef` should still happen.

slava77 replied:
uhm, on a second thought, perhaps not and the changes are OK, but that becomes a bit more obvious only after seeing that all cases to access the `dxy_` through `dtrkpt_` values rely on having `unpackVtx` called.

I see a path after this PR change that a candidate that comes out of this `PackedCandidate(const PackedCandidate &iOther)`  with `iOther.vertex_ == nullptr` can get to e.g. a call to `::packVtx` without a preceding `unpackVtx` (e.g. via `::setTrackProperties`) and lead to a crash in 
https://github.com/cms-sw/cmssw/blob/22f5521804d2e9a1ffee48aeefecca1bef226295/DataFormats/PatCandidates/src/PackedCandidate.cc#L26-L29

This looks like a `maybeUnpackBoth()` call is needed in `::setTrackProperties`

Dr15Jones replied:
Good catch! I made the change.

Dr15Jones commented:
> is there a chance that a ` PackedCandidate(PackedCandidate &&iOther)` is called or any of the assignment operators?

Those methods either already avoid reading the `vertex_` or only do so if `iOther.vertex_` is not null. Now they do read such values as `iOther.dphi_` which are a race condition if `iOther.vertex_ == nullptr`. However, those values (which are not guaranteed to be reasonable in that case) will never be used as the object will have `vertex_ == nullptr` which will trigger unpacking.

Dr15Jones committed:
Call maybeUnpackBoth from setTrackProperties

This protects the call to pack() which requires the variables
have already been unpacked.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32964/21238

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #32964 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10b87b/13064/summary.html
**COMMIT**: 85dfd14e120d154511fed208107ef135f1ba271a
**CMSSW**: CMSSW_11_3_X_2021-02-23-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/32964/13064/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 20 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2750983
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2750949
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/32964	85dfd14
- code changes are in line with the PR description and the follow up review; the physics outputs should not change
- jenkins tests pass and comparisons with the baseline show no relevant differences as expected
    - a few workflows like 140.56 have `ExcessiveTime` LogWarning in the PR tests for the same module in all 4 cases that I checked, with the message looking like 
```
%MSG-e ExcessiveTime:  SiPixelClusterProducer:siPixelClustersPreSplitting@cpu  24-Feb-2021 16:51:38 CET Run: 326479 Event: 1394020
ExcessiveTime: Module used 614.385 seconds of time which exceeds the error threshold configured in the Timing Service of 600 seconds.

```
As I recall in the past similar symptoms were an indication of some issues in the Frontier infrastructure

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
